### PR TITLE
Support Dynamo Tensor.backward by default

### LINKED ISSUE #167732: Dynamo should support backward() capture
### 🐛 Describe the bug

A prerequisite for this is https://github.com/pytorch/pytorch/issues/167729

Once you have torch.autograd.grad, the game is that you want to desugar backward() into a call to grad() plus updates to the grad fields on various tensors. The tricky part is that the only way to tell what parameters will be updated is by traversing the autograd graph, but you will need to be able to map the leaf variables to VariableTrackers when you do this. But this seems surmountable. The example in the other issue is setup to simulate backward().

### Versions

main

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185033

Tensor.backward already lowers through Dynamo's autograd.grad handler plus accumulate_grad_, but method_backward rejected the call whenever trace_autograd_ops was left at its default False. That meant normal torch.compile users still hit an Unsupported graph break even though the internal lowering path was present.

This removes the outer Tensor.backward gate and temporarily enables trace_autograd_ops only while issuing the synthetic autograd.grad call created by the backward lowering. Direct user calls to torch.autograd.grad still honor the public flag, and the existing autograd.grad safety checks continue to run for the lowered backward path.

A broad default flip of trace_autograd_ops was considered but prior CI showed it exposes too much user-facing autograd.grad tracing. Keeping the temporary config patch scoped to the internal backward rewrite fixes the requested behavior without changing direct autograd.grad default behavior.

Fixes #167732

Generated by my agent

Test Plan:

python test/dynamo/test_fwd_loss_bwd.py -k TestTensorBackwardDefaultConfig

python test/dynamo/test_fwd_loss_bwd.py -k tensor_backward

python test/dynamo/test_fwd_loss_bwd.py

lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the unclassified CI failures came from a resumed Dynamo frame silently treating backward as a no-op when its leaf tensors were no longer visible after a graph break. The update now falls back to eager backward for that case; the previously failing Dynamo-wrapped max-unpool and gumbel-softmax tests both pass locally, along with the full forward/loss/backward suite and lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the July 27 review and the related CI failures. Tensor.backward now falls back before lowering when post-accumulate-grad hooks or unsupported tensor/gradient layouts are present, guards real hook state so cached graphs invalidate correctly, fixes the split graph-break hint and missing path assertions, and removes discarded retry metrics that inflated HuggingFace graph-break counts. Sparse gather, jagged NestedTensor, complex convolution, the 50-test forward/loss/backward suite, the YituTechConvBert benchmark, and lintrunner pass locally; the branch is also rebased onto current main.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the July 31 hook review and the eight reported CI failures are addressed in this rebased update. Tensor.backward now bypasses trace_autograd_ops only for its explicitly marked synthetic autograd.grad call; direct calls remain gated. The lowering safely inspects retained GraphArgs, guards cache transitions, and falls back for pre-existing or in-graph Tensor hooks, post-accumulate hooks, direct AccumulateGrad node hooks, create_graph, wrapper subclasses, invalid explicit inputs, untracked leaves, and unsupported layouts. Restart accounting now rolls back only the exact discarded reason and owned counter occurrence under one lock. The 62-test forward/loss/backward file, rebuilt C++ binding, lintrunner, sparse/nested/max-unpool/gumbel/convolution and CPU op regressions all pass locally; YituTechConvBert improves to 4 graph breaks versus the expected 5. A fresh final review found no remaining blockers.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
